File: leetcode/validate-binary-search-tree/validate-binary-search-tree.cpp
/**************************************
* @author dujiong
* @date 2016.10.31
* @version V0.1
**************************************/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode *root) {
        if(root==NULL){
            return true;
        }
        getInorder(root);
        for(int i=0;i<inorder.size()-1;++i){
            if(inorder[i]>=inorder[i+1]){
                return false;
            }
        }
        return true;
    }
private:
    void getInorder(TreeNode* root){
        if(root==NULL){
            return;
        }
        getInorder(root->left);
        inorder.push_back(root->val);
        getInorder(root->right);
    }
    vector<int> inorder;
};

File: leetcode-golang/Validate Binary Search Tree/Validate Binary Search Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// [10,5,15,null,null,6,20]
 func isValidBST(root *TreeNode) bool {
    return isValidBSTHelper(root, nil, nil)
}

func isValidBSTHelper(root , l, r *TreeNode) bool {
	if root == nil {
		return true
	}
	if (l != nil && l.Val >= root.Val) || (r != nil && r.Val <= root.Val) {
		return false
	}
	return isValidBSTHelper(root.Left, l, root) && isValidBSTHelper(root.Right, root, r)
}